File: 238.java
import java.util.*;

class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] result = new int[n];
        Arrays.fill(result, 1);
        int last = 1;
        for (int i = 0; i < n; i++) {
            result[i] *= last;
            last *= nums[i];
        }
        last = 1;
        for (int i = n - 1; i >= 0; i--) {
            result[i] *= last;
            last *= nums[i];
        }
        return result;
    }
}

File: Python/238.py
from typing import List


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        result = [0] * len(nums)

        curr = 1
        for i, num in enumerate(nums):
            result[i] = curr
            curr *= num

        curr = 1
        for i, num in reversed(list(enumerate(nums))):
            result[i] *= curr
            curr *= nums[i]

        return result
